.github/workflows/main.yml

```diff
@@ -59,6 +63,27 @@ jobs:
         CI_DEPLOY_PASSWORD: ${{ secrets.CI_DEPLOY_PASSWORD }}
         # MAVEN_GPG_PASSPHRASE: ${{ secrets.MAVEN_GPG_PASSPHRASE }}
       run: ./mvnw -B -q -ff -DskipTests -ntp source:jar deploy
+    # Until generator_generic_slsa3@v1.2.3, calculate hashes from ./target
+    # https://github.com/slsa-framework/slsa-github-generator/issues/1225
+    - name: Generate hash
+      id: hash
+      if: |
+        github.event_name != 'pull_request' &&
+        matrix.java_version == '8' &&
+        endsWith(steps.projectVersion.outputs.version, '-SNAPSHOT')
```

Comment:
This would only calculate hashes for SNAPSHOT builds, not for actual releases?

Reply:
Ah ok.

So here we might have a challenge: releases are not triggered by CI -- I trigger them from my work station using Maven release plug-in (so second option).
But while I know how releases are often triggered via CI (from tagging), I haven't quite found those to my liking, at least the way most corporations do it: they tend to geared towards patch-increase only cases, for (web) applications where there is little semantic information in version numbers.

Work involved would also be quite substantial as well considering none of Jackson repos uses CI releases at this point.
And I do need to be able to release versions from multiple branches.

But longer term I can see why use of CI would be preferable to local work station -- for multiple reasons. So I am not philosophically against it at all. Just more concerned about practical path there.

One thing that might help would be finding good CI workflow to consider adopting: something used for library/framework use case where semantic versioning is used (or rather, where version number increments need to be somewhat more flexible; including occasional release of micro-patches).

I haven't really followed how other OSS projects deal with this so not sure where to look. I guess I could ask on Tidelift OSS slack channels; Joda project might be a good candidate to consider.

Reply:
Let me know if I'm missing something, but GitHub workflows should be able to handle different branches and be indifferent to tag naming conventions.

We could likely create a workflow very similar to this one but that's triggered:
- manually (`on: workflow_dispatch`). You'd pass the branch/tag/commit to use by hand. 
- whenever a new GitHub Release is created (`on: release`). So you define the branch/tag/commit from which to run. For an example (including SLSA provenance generation), see [the release workflow for the `sigstore-python` library](https://github.com/sigstore/sigstore-python/blob/main/.github/workflows/release.yml).

You can also use the `gh` CLI or simply `curl` to perform these actions remotely.

I'm not an expert on releasing on Maven, so I'm unsure if there are additional steps required for actual releases.

(Let me know if you'd rather migrate this conversation somewhere more appropriate)

Reply:
> And I do need to be able to release versions from multiple branches.

It just occurred to me you might've meant you need to be able to release from multiple repos (jackson & jackson-core, for example). In this case, we could write a shell script to make the `gh/curl` calls for all the repos at once.

Reply:
My concern is not that Github workflows can't do things, but rather how awkward their use is. Right now use of

    mvn release:clean release:prepare release:perform

is straight-forward with little git interaction needed. Increases versions etc, handles everything except for final approval on Sonatype OSS repo (which I'll need to do anyway).
Pushing tags etc has always seemed pretty clunky in comparison.

And wrt multi-repo release, there is no workflows for that so no need to worry about that yet.

Reply:
Sorry for the delay in replying, I've been getting more familiar with the Maven release plugin.

As I've come to understand it from [the docs](https://maven.apache.org/maven-release/maven-release-plugin/examples/prepare-release.html), the magic really happens in the `prepare` stage. My understanding is that `prepare`:

- Bumps everything from x-SNAPSHOT to the new release version
- Commits the changed POMs
- Creates a tag for the new version
- Bumps everything to the next y-SNAPSHOT
- Commits the changed POMs again

Meanwhile, `perform` simply builds and deploys. From a security standpoint, however, the most important step to have in CI is actually `perform`.

I've therefore had the idea of splitting the workflow into two parts: we can write a shell script you run on your local machine that runs `:clean :prepare` and pushes everything to GitHub (including the tag created by `prepare`). We can then have a GitHub workflow that triggers on new tags (excluding `*-rc` release-candidate tags) and simply runs the final `:perform`.

Your release process would therefore be almost unchanged. Instead of calling `mvn release:clean release:prepare release:perform`, you'd call `release.sh`, which would give you the same behavior as `mvn release:clean release:prepare` and then automatically trigger the CI `mvn release:perform` by pushing the new commits and tag.

Would you be interested in exploring this?

Reply:
Hi @pnacht ! Apologies for slow response here. But: yes, I think that sounds like a reasonable compromise and I would be interested.
I think we could play with this for core components (jackson-core, jackson-databind, jackson-annotations).

One quick note: I will be now off for vacation until January 4th 2023 so let's sync up early next year.

Reply:
Hey @cowtowncoder, hope you had a good vacation! I've started sketching out the script to replace `mvn release:clean release:prepare` and should have the workflow that'll do the final `mvn release:perform` end of next week. I'll submit those as a separate PR, is that alright?

Reply:
@pnacht That makes sense to me (separate PR)

Reply:
Sorry for the delay! But I've submitted the PR as discussed above: https://github.com/FasterXML/jackson-core/pull/896.